Optimizes interactive scene for homogeneous cloning

kellyguo11 opened the pull request:
# Description

Previously, we always assumed heterogeneous environment setup, which performed cloning on a per-actor basis during spawning of actors. In some direct workflow environments, a redundant cloning call was made at the end of environment setup to perform cloning on a global scope. 

With this change, we determine whether to perform cloning on a per-actor or global scope based on the `replicate_physics` parameter in `InteractiveSceneCfg`. When `replicate_physics=True`, homogeneous environment is assumed and cloning will only happen on a global scope. This reduces the time required for cloning for homogeneous setups. For heterogeneous environments, set `replicate_physics=False` and cloning will happen per-actor.

## Type of change

<!-- As you go through the list, delete the ones that are not applicable. -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update


## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/scene/interactive_scene_cfg.py:
```diff
@@ -85,4 +85,8 @@ class MySceneCfg(InteractiveSceneCfg):
-    """Enable/disable replication of physics schemas when using the Cloner APIs. Default is True."""
+    """Enable/disable replication of physics schemas when using the Cloner APIs. Default is True.
+
+    If true, the environment will be assumed to be homogeneous and cloning happens on a global scope.
+    Set this to False for heterogeneous environments, where cloning happens per-object.
```
Homogenous and hetergoneous cloning might be less intuitive?

Also we should document this in the Scene class docstrings as well. Also it seems that the documentation on usage there is limited. Can we fix that too?

```suggestion
    If true, the simulation will have the same asset instances (USD prims) in all the cloned environments. Internally, this ensures optimization
    in setting up the scene and parsing it via the physics stage parser.
    
    If False, the simulation allows having separate asset instances (USD prims) in each environment. This flexibility comes
    at a cost of slowdowns in setting up and parsing the scene.
    
    .. note::
        Optimized parsing of certain prim types (such as deformable objects) is not currently supported
        by the physics engine. In these cases, this flag needs to be set to False.
```

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/scene/interactive_scene.py:
```diff
@@ -128,12 +136,14 @@ def clone_environments(self, copy_from_source: bool = False):
             replicate_physics=self.cfg.replicate_physics,
             copy_from_source=copy_from_source,
         )
+        if self._default_env_origins is None:
```
```suggestion
       # in case of heterogenous cloning, the env origins is specified at init
        if self._default_env_origins is None:
```

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/scene/interactive_scene.py:
```diff
@@ -90,20 +90,28 @@ def __init__(self, cfg: InteractiveSceneCfg):
         self.cloner = GridCloner(spacing=self.cfg.env_spacing)
         self.cloner.define_base_env(self.env_ns)
         self.env_prim_paths = self.cloner.generate_paths(f"{self.env_ns}/env", self.cfg.num_envs)
+        # environment origins
```
would move this to an if-else at L-99 for readability sake

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/scene/interactive_scene.py:
```diff
@@ -90,20 +90,28 @@ def __init__(self, cfg: InteractiveSceneCfg):
+                copy_from_source=True,
+            )
+            self._default_env_origins = torch.tensor(env_origins, device=self.device, dtype=torch.float32)
         self._global_prim_paths = list()
```
When user calls `filter_collisions(global_prim_paths)` should we store the global prim paths from that call into this list?

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/scene/interactive_scene.py:
```diff
@@ -128,12 +136,14 @@ def clone_environments(self, copy_from_source: bool = False):
+        if self._default_env_origins is None:
+            self._default_env_origins = torch.tensor(env_origins, device=self.device, dtype=torch.float32)
 
     def filter_collisions(self, global_prim_paths: list[str] = []):
```
Not related to this MR but `[]` in the args is not a great practice since it is a mutable data type.

https://docs.aws.amazon.com/codeguru/detector-library/python/default-argument-mutable-objects/

kellyguo11 commented:
Addressed the comments.

For the list argument in `def filter_collisions(self, global_prim_paths: list[str] = []):`, I think this one will be a breaking change. If we go for it, perhaps we should do it in a new MR?

Mayankm96 commented:
> For the list argument in def filter_collisions(self, global_prim_paths: list[str] = []):, I think this one will be a breaking change. If we go for it, perhaps we should do it in a new MR?

I don't think that is a breaking API change since it was a default argument. I don't think anyone was calling filter_collisions multiple times so far so we should be able to do this change without any breaking of the API itself.

Though if you think it is better, we can make a separate MR for it. We should fix it though :)

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/scene/interactive_scene.py:
```diff
@@ -155,6 +194,10 @@ def filter_collisions(self, global_prim_paths: list[str] = []):
             global_paths=global_prim_paths,
         )
 
+        # set global prim paths list if not previously defined
+        if len(self._global_prim_paths) < 1:
+            self._global_prim_paths += global_prim_paths
```
Should this happen before in the function and we should give self._global_prim_paths to filter collision flags?

I guess this will support the case where people do a mix of using scene and their own spawning (though there's no reason why they should). We should also remove duplicate elements from the list or check for them?

Mayankm96 replied:
If this goes outside the MR scope, let's do it properly in a separate MR like you suggested.

kellyguo11 replied:
Updated.

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/scene/interactive_scene.py:
```diff
@@ -174,30 +174,37 @@ def clone_environments(self, copy_from_source: bool = False):
 
         Args:
-            global_prim_paths: The global prim paths to enable collisions with.
+            global_prim_paths: A list of global prim paths to enable collisions with.
```
```suggestion
            global_prim_paths: A list of global prim paths to enable collisions with.
                Defaults to None, in which case no global prim paths are considered.
```
